paddle/fluid/operators/pad3d_op.cc

```diff
@@ -0,0 +1,1011 @@
+          }
+        }
+      }
+    }
+    d_in_data += in_depth * in_height * in_width * channels;
+    d_out_data += out_depth * out_height * out_width * channels;
+  }
+}
+
+template <typename T>
+void Pad3DGradCircularNCDHW(T* d_in_data, const int num, const int channels,
+                            const int in_depth, const int in_height,
+                            const int in_width, const int out_depth,
+                            const int out_height, const int out_width,
+                            const int pad_front, const int pad_top,
+                            const int pad_left, const T* d_out_data) {
+  for (int n = 0; n < num; ++n) {
+    for (int c = 0; c < channels; ++c) {
+      for (int out_d = 0; out_d < out_depth; ++out_d) {
+        for (int out_h = 0; out_h < out_height; ++out_h) {
+          for (int out_w = 0; out_w < out_width; ++out_w) {
+            int in_d = ((out_d - pad_front) % in_depth + in_depth) % in_depth;
+            int in_h = ((out_h - pad_top) % in_height + in_height) % in_height;
+            int in_w = ((out_w - pad_left) % in_width + in_width) % in_width;
+            d_in_data[in_d * in_height * in_width + in_h * in_width + in_w] +=
+                d_out_data[out_d * out_height * out_width + out_h * out_width +
+                           out_w];
+          }
+        }
+      }
+      d_in_data += in_depth * in_height * in_width;
+      d_out_data += out_depth * out_height * out_width;
+    }
+  }
+}
+
+template <typename T>
+void Pad3DGradCircularNDHWC(T* d_in_data, const int num, const int channels,
+                            const int in_depth, const int in_height,
+                            const int in_width, const int out_depth,
+                            const int out_height, const int out_width,
+                            const int pad_front, const int pad_top,
+                            const int pad_left, const T* d_out_data) {
+  for (int n = 0; n < num; ++n) {
+    for (int out_d = 0; out_d < out_depth; ++out_d) {
+      for (int out_h = 0; out_h < out_height; ++out_h) {
+        for (int out_w = 0; out_w < out_width; ++out_w) {
+          int in_d = ((out_d - pad_front) % in_depth + in_depth) % in_depth;
+          int in_h = ((out_h - pad_top) % in_height + in_height) % in_height;
+          int in_w = ((out_w - pad_left) % in_width + in_width) % in_width;
+
+          const int out_index =
+              (out_d * out_height * out_width + out_h * out_width + out_w) *
+              channels;
+          const int in_index =
+              (in_d * in_height * in_width + in_h * in_width + in_w) * channels;
+          for (int c = 0; c < channels; ++c) {
+            d_in_data[in_index + c] += d_out_data[out_index + c];
+          }
+        }
+      }
+    }
+    d_in_data += in_depth * in_height * in_width * channels;
+    d_out_data += out_depth * out_height * out_width * channels;
+  }
+}
+
+static inline void GetPaddings(int* paddings,
```

Comment:
simply return const int pointer to tensor.data or vector.data as`pads` are read only?

Reply:
done, thx